runners/direct-java/src/test/java/org/apache/beam/runners/direct/BoundedReadEvaluatorFactoryTest.java

```diff
@@ -434,9 +432,13 @@ public boolean startImpl() throws IOException {
 
     @Override
     public boolean advanceImpl() throws IOException {
-      if (index == sleepIndex) {
+      // Sleep before the sleep/split index is claimed so long as it will be claimed
+      if (index + 1 == sleepIndex && sleepIndex < getCurrentSource().elems.length) {
         try {
           dynamicallySplit.await();
+          while (initialSource.equals(getCurrentSource())) {
+            // Spin until the current source is updated
```

Comment:
`OffsetBasedSource#splitAtFraction` is final, so we can't do it there. There also isn't any reader method called after the source is updated, so we can't add a hook there.

I've added a sleep of 1ms to make sure the CPU is yielded here.